TLM/TLM/State/SavedGameOptions.cs

```diff
@@ -0,0 +1,136 @@
+    public bool highwayRules;
+    public bool highwayMergingRules;
+    public bool automaticallyAddTrafficLightsIfApplicable = true;
+    public bool NoDoubleCrossings;
+    public bool DedicatedTurningLanes;
+
+    public bool showLanes = VersionUtil.IS_DEBUG;
+
+    public bool strongerRoadConditionEffects;
+    public bool parkingAI;
+    public bool disableDespawning;
+    public bool preferOuterLane;
+    //public byte publicTransportUsage = 1;
+
+    public bool prioritySignsEnabled = true;
+    public bool timedLightsEnabled = true;
+    public bool customSpeedLimitsEnabled = true;
+    public bool vehicleRestrictionsEnabled = true;
+    public bool parkingRestrictionsEnabled = true;
+    public bool junctionRestrictionsEnabled = true;
+    public bool turnOnRedEnabled = true;
+    public bool laneConnectorEnabled = true;
+
+    public VehicleRestrictionsAggression vehicleRestrictionsAggression = VehicleRestrictionsAggression.Medium;
+    public bool RoundAboutQuickFix_DedicatedExitLanes = true;
+    public bool RoundAboutQuickFix_StayInLaneMainR = true;
+    public bool RoundAboutQuickFix_StayInLaneNearRabout = true;
+    public bool RoundAboutQuickFix_NoCrossMainR = true;
+    public bool RoundAboutQuickFix_NoCrossYieldR = false;
+    public bool RoundAboutQuickFix_PrioritySigns = true;
+    public bool RoundAboutQuickFix_KeepClearYieldR = true;
+    public bool RoundAboutQuickFix_RealisticSpeedLimits;
+    public bool RoundAboutQuickFix_ParkingBanMainR = true;
+    public bool RoundAboutQuickFix_ParkingBanYieldR;
+    public bool PriorityRoad_CrossMainR;
+    public bool PriorityRoad_AllowLeftTurns;
+    public bool PriorityRoad_EnterBlockedYeild;
+    public bool PriorityRoad_StopAtEntry;
+
+    // See PathfinderUpdates.cs
+    public byte SavegamePathfinderEdition = PathfinderUpdates.LatestPathfinderEdition;
+
+    public bool showDefaultSpeedSubIcon;
+
+    internal int getRecklessDriverModulo() => CalculateRecklessDriverModulo(recklessDrivers);
+
+    internal static int CalculateRecklessDriverModulo(RecklessDrivers level) => level switch {
+        RecklessDrivers.PathOfEvil => 10,
+        RecklessDrivers.RushHour => 20,
+        RecklessDrivers.MinorComplaints => 50,
+        RecklessDrivers.HolyCity => 10000,
+        _ => 10000,
+    };
+
+    /// <summary>
+    /// Determines whether Dynamic Lane Selection (DLS) is enabled.
+    /// </summary>
+    public bool IsDynamicLaneSelectionActive() {
+        return advancedAI && altLaneSelectionRatio > 0;
+    }
+
+    /// <summary>
+    /// When <c>true</c>, options are safe to query.
+    /// </summary>
+    /// <remarks>
+    /// Is set <c>true</c> after options are loaded via <see cref="Manager.Impl.OptionsManager"/>.
+    /// Is set <c>false</c> while options are being loaded, and also when level unloads.
+    /// </remarks>
+    public static bool Available { get; set; } = false;
+
+    public static SavedGameOptions Instance { get; private set; }
```

Comment:
this is inlined. I proved using this patch 
![image](https://user-images.githubusercontent.com/26344691/204596058-faeab9a5-70c3-444f-8fe0-1a7b433c584b.png)
No such log.

Reply:
Are you use that Prefix didn't get inlined and basically harmony path was not even created? 😅 
Wrap that log with try...catch to prevent inlining

Reply:
I don't think prefix inlining can ever been an issue.

Reply:
I've seen in multiple mods that authors were decorating it with NoInlining attribute, especially if method was a placeholder for harmony reverse patch they add dummy try..catch inside since using that structure in the method body automatically prevents it from inlining.
Anyways, I didn't test if that's the case here, just pointing out possible issue/unexpected behavior in the test itself.

Reply:
even aggressive in-lining does not work. @krzychu124 are you sure `MethodImpl(256)` is supposed to work?
```cs
    public static SavedGameOptions Instance {
        [MethodImpl(Constants.AggressiveInlining)]
        get;
        private set;
    }
```
```
Info 181.1571230: SavedGameOptions.Instance is NOT inlined
```

Reply:
> Windows10CE
do note you need to try and access the property before you patch
MonoMod is sometimes able to disable inlining when it patches a method, so if you're only testing the property after patching but not before you might be hitting that fix

Reply:
`Test` is never in the stack trace which proves without the patch its inlined even without `MethodImpl(256)`
```cs
    public static void Test() {
            Log.Info($"SavedGameOptions.Test() called : Instance.advancedAI = {Instance?.advancedAI}");
    }
```
```
Info 464.8905046: SavedGameOptions.Instance is NOT inlined
   at System.Environment.get_StackTrace()
   at TrafficManager.State.SavedGameOptionsPatch.Prefix()
   at TrafficManager.State.SavedGameOptions.TrafficManager.State.SavedGameOptions.get_Instance_Patch1()
   at TrafficManager.Patch._RoadBaseAI.GetTrafficLightNodeStatePatch.Prefix(UInt16 nodeID, NetNode ByRef nodeData, UInt16 segmentID, NetSegment ByRef segmentData, Flags ByRef flags, Color ByRef color)
   at RoadBaseAI.RoadBaseAI.GetTrafficLightNodeState_Patch1(UInt16 , NetNode ByRef , UInt16 , NetSegment ByRef , FlagsLong ByRef , Color ByRef )
   at RoadBaseAI.GetNodeState(UInt16 nodeID, NetNode ByRef nodeData, UInt16 segmentID, NetSegment ByRef segmentData, FlagsLong ByRef flags, Color ByRef color)
   at NetSegment.RenderInstance(.CameraInfo cameraInfo, UInt16 segmentID, Int32 layerMask, .NetInfo info, Instance ByRef data)
   at NetSegment.RenderInstance(.CameraInfo cameraInfo, UInt16 segmentID, Int32 layerMask)
   at NetManager.EndRenderingImpl(.CameraInfo cameraInfo)
   at SimulationManagerBase`2.EndRendering(.CameraInfo cameraInfo)
   at NetManager.IRenderableManager.EndRendering(.CameraInfo cameraInfo)
   at RenderManager.LateUpdate()
Info 464.8938528: SavedGameOptions.Test() called : Instance.advancedAI = False
Info 464.9042487: SavedGameOptions.Instance is NOT inlined
   at System.Environment.get_StackTrace()
   at TrafficManager.State.SavedGameOptionsPatch.Prefix()
   at TrafficManager.State.SavedGameOptions.TrafficManager.State.SavedGameOptions.get_Instance_Patch1()
   at TrafficManager.Patch._RoadBaseAI.GetTrafficLightNodeStatePatch.Prefix(UInt16 nodeID, NetNode ByRef nodeData, UInt16 segmentID, NetSegment ByRef segmentData, Flags ByRef flags, Color ByRef color)
   at RoadBaseAI.RoadBaseAI.GetTrafficLightNodeState_Patch1(UInt16 , NetNode ByRef , UInt16 , NetSegment ByRef , FlagsLong ByRef , Color ByRef )
   at RoadBaseAI.GetNodeState(UInt16 nodeID, NetNode ByRef nodeData, UInt16 segmentID, NetSegment ByRef segmentData, FlagsLong ByRef flags, Color ByRef color)
   at NetSegment.RenderInstance(.CameraInfo cameraInfo, UInt16 segmentID, Int32 layerMask, .NetInfo info, Instance ByRef data)
   at NetSegment.RenderInstance(.CameraInfo cameraInfo, UInt16 segmentID, Int32 layerMask)
   at NetManager.EndRenderingImpl(.CameraInfo cameraInfo)
   at SimulationManagerBase`2.EndRendering(.CameraInfo cameraInfo)
   at NetManager.IRenderableManager.EndRendering(.CameraInfo cameraInfo)
   at RenderManager.LateUpdate()
   ```

Reply:
> even aggressive in-lining does not work. @krzychu124 are you sure `MethodImpl(256)` is supposed to work?
> 
> ```cs
>     public static SavedGameOptions Instance {
>         [MethodImpl(Constants.AggressiveInlining)]
>         get;
>         private set;
>     }
> ```
> 
> ```
> Info 181.1571230: SavedGameOptions.Instance is NOT inlined
> ```

I've never said you can force inline. That is impossible, you can only suggest it. Honestly I'm not sure what you expect. Inline process replaces method call with instructions from the method body the call is triggered. In one method it might be inlined, in the other not. 
If you are expecting to **not see** that log call (entry _"SavedGameOptions.Instance is NOT inlined"_) then I can assure you it won't ever happen. _Text will be logged, always!_

Reply:
> only sometimes is inlined.

I don't agree with your conclusion. Our experiment is influencing the outcome. I think its always inlined (without the patch). its inlined when JITed if it is JITed before the patch. Those places where it is JITed after the patch, inlining won't happen. But if you remove the patch then I think inlining would always happen. To prove that we need to force all codes that calls Instance to be JITed BEFORE the patch is applied. but like you said we have more fun things to do.

Reply:
Want more experiments?
Create static method `Test2` that calls your static `Test`. It `may` appear in the stacktrace even if you think `its always inlined`.
I did few other experiments, the thing is... inline more likely happen for **static methods** from quite obvious reason: by definition they won't change and don't depend on any instance, so their code can be safely injected instead of calling the method but the instruction count matters (something you can't really see since `IL body != jitted method`). 
That's not the case for instance methods.

Reply:
I suppose we have to agree to disagree. In any case I did cache savedGameOptions instance just like global config. so lets PR approved so that I can continue with XML and saved as default for new games.